src/lib/storage/chunk.cpp

```diff
@@ -297,4 +297,31 @@ void Chunk::set_cleanup_commit_id(const CommitID cleanup_commit_id) {
   _cleanup_commit_id.store(cleanup_commit_id);
 }
 
+void Chunk::mark_as_finalizable() {
+  Assert(!_is_finalizable, "Chunk should not be marked as finalizable multiple times.");
+  _is_finalizable = true;
+}
+
+void Chunk::try_finalize() {
+  DebugAssert(_mvcc_data, "Expected to be executed with MVCC enabled.");
+  // Finalize if the chunk was marked finalizable (i.e., it reached the target size and a new chunk was added to the
+  // table), it is still mutable and all pending Insert operators are either commited or rolled back. We do not have to
+  // set the max_begin_cid here since committed Insert operators already set it.
+  if (!_is_finalizable || !is_mutable() || _mvcc_data->pending_inserts() != 0) {
+    return;
+  }
+
+  // Mark chunk as immutable. fetch_and() is only defined for integral types, so we use compare_exchange_strong().
+  auto success = true;
+  if (_is_mutable.compare_exchange_strong(success, false)) {
+    DebugAssert(success, "Value exchanged but value was actually false.");
+    // We were the first ones to mark the chunk as immutable. Thus, we have to take care of anything else that needs to
+    // be done. In the future, this can mean to start background statistics generation, encoding, etc.
+    _is_finalizable = false;
```

Comment:
We use the atomic compare/exchange to make sure we (and only we) are in charge of finalizing the chunk. This does not really work with `finalize()`.

Reply:
To me, the interface appears confusing. When the function is called `try_finalize()`, it should finalize when possible.

Also, it very carefully sets `_is_mutable` while `finalize()` just sets it. Might we want to get rid of `finalize()`?

Reply:
I mean, that's pretty much what's happening. `try_finalize()` finalizes whenever possible, e.g., the chunk is marked as finalizable (i.e., it reached its target size and another mutable chunk was appended tot the table) and all pending Insert operators finished. And finalizing per now is just marking the chunk as immutable, correct.

`finalize()`, however, does not care about these two prerequisites and simply marks the chunk as immutable. I think it's fair to have both so we can finalize in test cases and when loading data without having to allow the chunk to be finalized (or to forbid it when the chunk was appended by an Insert operator).

Thus, `finalize()` also tries to set the min/max CommitID as it is the "no-transaction" version. If you like to, we could merge both to `try_finalize()` and add a function to forbid the chunk to be finalized, which is then called by the Insert operator appending the chunk. But having both functions sounds fine for me.

Reply:
Maybe the other way around? `try_finalize()` does not call `finalize()`, but `finalize()` marks the chunk as finalizable and calls `try_finalize()`? We could also shift setting the max begin CID in `try_finalize()` if required. However, this could have unwanted side effects if we want to autmatically spin up further encoding/statistics generation for chunks when the system is running.

Reply:
Not doing the MVCC stuff when we are in the situation we have guaranteed only one actor (the one who is finalizing) and we are sure that it is safe, would be kind of a waste.
Why would this interfere with encoding?

Reply:
I would love to have a white board right now ...

We have "two kinds" auto finalize, right? The "new" automated one triggered through inserts/updates and the ones when somebody wants to finalize a chunk that is no longer written (e.g., our benchmarks).

Why do we need both functions? Can't we have only `try_finalize()` that returns a bool? This one would set the MVCC stuff.
Encoding and statistics are entirely independent from everything here, aren't they? (apart from the constraint of chunks being immutable)

Reply:
I beg to differ. As `try_finalize()` is called in the transaction world, all the MVCC stuff _is_ already set. We could try to set it here, but in fact, we are guaranteed it exists.

Do you siggest that everyone calling `try_finalize()` then (i.e., the Insert operator) and gets a `true` spins up encoding etc? Doesn't sound too appealing to me tbh. I mean, we have no semantics what "finalizing" a chunk actually means. Is it only marking as immutable? Is it encoding and statistics generation? That's up to us. I would say the version called during transactions is the exact entry point for doing that, but we might not want this to be done when loading data or testing.

What I could live with is the following: Rename `finalize()` to `mark_as_immutable()` as it does that (plis the MVCC stuff maybe). Then, `try_finalize()` can do more stuff in the future. Maybe we can move the careful marking to the first and call it from the second or wrap it elsewhere so we can ensure the calls from the non-transaction side always succeed. Does this sound like a comprimise?

Reply:
> Do you siggest that everyone calling try_finalize() then (i.e., the Insert operator) and gets a true spins up encoding etc?

No, definitely not.
I think we are pretty much on the same track, but it might make sense to have a quick chat about that. As stated above, I don't like `finalize()`, just as you said.
We need a white board (well, I need one to understand everything here :D).